### PR TITLE
Respect user set Type for client errors via ProducesResponseType

### LINKED ISSUE #7874: ProducesResponseType(typeof(void), ...) causes ApiExplorer to indicate the return type is ProblemDetails
### Describe the bug
ApiExplorer/ProducesResponseType exposes incorrect value when the return type is void.  It seems to override any attempt to indicate that the response produces no results/body with ProblemDetails.

### To Reproduce
1. Using this version of ASP.NET Core 2.2
2. Apply the following attribute to an ApiController:
[ProducesResponseType(typeof(void), (int)HttpStatusCode.Unauthorized)
3. ApiExplorer indicates the return type is ProblemDetails.

### Expected behavior
ApiExplorer should say the return type is void.

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/7874.

Follow-up to https://github.com/dotnet/aspnetcore/pull/30236.

### CODE REVIEW

File: src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs [hunk 1]
```diff
@@ -101,6 +101,7 @@ private static List<IApiResponseMetadataProvider> GetResponseMetadataAttributes(
 
                     if (apiResponseType.Type == typeof(void))
                     {
+                        var setByDefault = metadataAttribute is ProducesResponseTypeAttribute attribute && attribute.SetByDefault;
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs:104]:
```suggestion
                        var setByDefault = metadataAttribute is ProducesResponseTypeAttribute { SetByDefault: true };
```

for 💯

Comment (pranavkm) [reply src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs:104]: Might help to add a comment about this. 

> We're trying to differentiate between a `typeof(void)` that is implicitly assigned by `ProduceResponseTypeAttribute` vs one that is explicitly assigned.

File: src/Mvc/Mvc.Core/src/ProducesResponseTypeAttribute.cs [hunk 2]
```diff
@@ -43,6 +45,8 @@ public ProducesResponseTypeAttribute(Type type, int statusCode)
         /// </summary>
         public int StatusCode { get; set; }
 
+        internal bool SetByDefault { get; private set; }
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/ProducesResponseTypeAttribute.cs:48]: You don't need to document `internal` apis, but it never hurts to add one. Also, `IsResponseTypeImplicitlyAssigned` or `IsResponseTypeSetByDefault`

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/ProducesResponseTypeAttribute.cs:48]:
```suggestion
        internal bool SetByDefault { get; }
```

File: src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs (hunk 1 above)
Comment (captainsafia) [reply src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs:104]: #pranavpoints 

Did I use that right?

**Review by pranavkm — APPROVED** [review]

**Review by javiercn — APPROVED** [review]
Looks great!

**Pull request merged**